responses/__init__.py

```diff
@@ -640,9 +640,10 @@ def replace(self, method_or_response=None, url=None, body="", *args, **kwargs):
         else:
             response = Response(method=method_or_response, url=url, body=body, **kwargs)
 
-        try:
-            index = self._matches.index(response)
-        except ValueError:
+        for index, match in self._matching_sequence():
```

Comment:
If this now sometimes loops in reverse isn't the doc string inaccurate as the first response isn't necessarily the one that will be replaced.

Reply:
Kind of inaccurate, because now we would need to differentiate between *first response* in insertion order and *first response* in matching order. I think that the matching order is more relevant for users in most cases.

This leads me to a question about the intended use of the public `registered` method.

The documentation says:
> Registered responses are available as a public method of the RequestMock instance. It is sometimes useful for debugging purposes to view the stack of registered responses which can be accessed via responses.registered().

Currently, the `registered` method gives direct access to the **mutable private attribute** `_matches`, which allows direct manipulation of the internal state of the `RequestsMock` instance. I know that there is not a list view object in python, but from the documentation, it appears to me like the user should only get a view by this method and manipulate the registry with the `add`, `replace`, `upsert`, and `remove` methods.
Such a view (even though static compared to `dict.values()` for instance) could be implemented by returning `_matches.copy()`, or `[response for index, response in self._matching_sequence()]`. I would vote for the second one, because as said I think matching order is more relevant to the user. I would also say that it is most backward compatible for the debugging use-case mentioned in the documentation, meaning that users would like to know why another match had been picked instead of the one they expected, so they want to see the matching order before and after doing their request.
Now the question: do you see any conflict or critical backward incompatibility by changing the return value of the `registered` method?

Reply:
> Currently, the registered method gives direct access to the mutable private attribute _matches, which allows direct manipulation of the internal state of the RequestsMock instance. I

I think giving a mutable reference isn't the end of the world given that this library's primary purpose is to help with testing. We could clone the list but that adds complexity and potential for breaking userland code where we don't need to incur the complexity or risk.